### ISSUE TITLE
[data] `Dataset.stats()` causes `Dataset.schema()` to be unknown

### ISSUE THREAD

**Issue by matthewdeng** [ray-project/ray#29592]
### What happened + What you expected to happen

Calling `Dataset.stats()` will change the Dataset's schema to be `None`.

In my workload, I can workaround this by simply removing the  `stats()` call. However, the severity of this may increase depending on the underlying root cause, and if other methods will trigger the same code path in the execution plan.

This is problematic because some downstreams may depend on the schema (e.g. to determine default batch format, or to validate column names).

### Versions / Dependencies

Ray Nightly

### Reproduction script

```python
import ray

ds = ray.data.read_csv("s3://anonymous@air-example-data/breast_cancer.csv")
print(f"Dataset(Before): {ds}\n")
print(f"Stats: {ds.stats()}")
print(f"Dataset(After): {ds}")
```

```
Dataset(Before): Dataset(num_blocks=1, num_rows=569, schema={mean radius: double, mean texture: double, mean perimeter: double, mean area: double, mean smoothness: double, mean compactness: double, mean concavity: double, mean concave points: double, mean symmetry: double, mean fractal dimension: double, radius error: double, texture error: double, perimeter error: double, area error: double, smoothness error: double, compactness error: double, concavity error: double, concave points error: double, symmetry error: double, fractal dimension error: double, worst radius: double, worst texture: double, worst perimeter: double, worst area: double, worst smoothness: double, worst compactness: double, worst concavity: double, worst concave points: double, worst symmetry: double, worst fractal dimension: double, target: int64})

Stats: Stage 0 read: 1/1 blocks executed in 1.82s
* Remote wall time: 286.9ms min, 286.9ms max, 286.9ms mean, 286.9ms total
* Remote cpu time: 213.6ms min, 213.6ms max, 213.6ms mean, 213.6ms total
* Peak heap memory usage (MiB): 153.64 min, 153.64 max, 153 mean
* Output num rows: 569 min, 569 max, 569 mean, 569 total
* Output size bytes: 141112 min, 141112 max, 141112 mean, 141112 total
* Tasks per node: 1 min, 1 max, 1 mean; 1 nodes used

Dataset(After): Dataset(num_blocks=1, num_rows=569, schema=Unknown schema)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by matthewdeng** [comment]
cc @c21

**Comment by jianoaix** [comment]
I took a look. I think it's because we null out the schema at stats actor: https://sourcegraph.com/github.com/ray-project/ray@master/-/blob/python/ray/data/_internal/stats.py?L125

**Comment by c21** [comment]
Thanks @jianoaix for taking a look. Assigning issue to you for now.

**Comment by jianoaix** [comment]
Actually this script seems not able to reproduce for me (run "python script.py"). @matthewdeng

**Comment by matthewdeng** [comment]
[Here](https://console.anyscale-staging.com/o/anyscale-internal/workspaces/expwrk_PCM3KgMKjKV6TYmuJn3F8s5h/ses_m6kr6sQdYhJQjw7pmPrkTyUj)'s a Workspace where I'm able to reproduce the issue. Maybe it has to do with multi-node?

Ray commit is 2189552ada299a23c8dabf49ef10f4a9ecad68ca.

**Comment by jianoaix** [comment]
Hmm, I can repro in python interactive mode with the exact same script (run them one by one), but cannot if running it with "python script.py" :exploding_head:

**Comment by richardliaw** [comment]
awesome! Can you file a patch then?

**Comment by matthewdeng** [comment]
Fixed by #29635.
